test/Test/unit.jl

```diff
@@ -14,6 +14,7 @@ end
     # `UniversalFallback` needed for `MOI.Silent`
     mock = MOIU.MockOptimizer(MOIU.UniversalFallback(MOIU.Model{Float64}()))
     MOI.set(mock, MOI.Silent(), true)
+    MOI.set(mock, MOI.TimeLimitSec(), nothing)
```

Comment:
This is the default value, so why is the `set` needed?

Reply:
I think it should be remove. My bad.

Reply:
Well, removing that line breaks. `nothing` is the default value but I didn't set it in first place. I put that line back for now.
I don't know how to set a default value, and I think `Silent()` default value isn't set too. I'll be looking into it soon.

Reply:
What breaks? This needs to be resolved before we can merge.

Reply:
When `get()` is used on `TimeLimitSec()` instead of returning `nothing` as default value it throws an error. That happened in `src/Test/UnitTests/attributes.jl`:
`KeyError: key MathOptInterface.TimeLimitSec() not found`

Reply:
```Julia
(MathOptInterface) pkg> test MathOptInterface
   Testing MathOptInterface
 Resolving package versions...
Test Summary: | Pass  Total
MOI           |  116    116
Test Summary: | Pass  Total
MOI.Utilities | 5220   5220
time_limit_sec: Error During Test at /.../MathOptInterface.jl/src/Test/config.jl:45
  Got exception outside of a @test
  KeyError: key MathOptInterface.TimeLimitSec() not found
  Stacktrace:
   ...
time_limit_sec: Error During Test at /.../MathOptInterface.jl/src/Test/config.jl:45
  Got exception outside of a @test
  KeyError: key MathOptInterface.TimeLimitSec() not found
  Stacktrace:
   ...
Test Summary:                             | Pass  Error  Total
MOI.Test                                  | 3197      2   3199
  Config                                  |    6             6
  Unit                                    | 1062      2   1064
    Basic Constraint Tests                |  582           582
    Unit Tests                            |  343      2    345
      max_sense                           |    3             3
      min_sense                           |    3             3
      get_objective_function              |    5             5
      silent                              |    4             4
      time_limit_sec                      |    1      1      2
      getvariable                         |    2             2
      feasibility_sense                   |    3             3
      solver_name                         |    1             1
      delete_variables                    |    4             4
      getconstraint                       |    7             7
      variablenames                       |    5             5
      add_variables                       |    3             3
      delete_variable                     |    4             4
      add_variable                        |    3             3
      max_sense                           |    3             3
      min_sense                           |    3             3
      get_objective_function              |    5             5
      silent                              |    4             4
      time_limit_sec                      |    1      1      2
      getvariable                         |    2             2
      feasibility_sense                   |    3             3
      solver_name                         |    1             1
      delete_variables                    |    4             4
      getconstraint                       |    7             7
      variablenames                       |    5             5
      add_variables                       |    3             3
      delete_variable                     |    4             4
      add_variable                        |    3             3
      solve_blank_obj                     |   18            18
      solve_constant_obj                  |   10            10
      solve_singlevariable_obj            |   10            10
      solve_with_lowerbound               |   13            13
      solve_with_upperbound               |   13            13
      solve_affine_lessthan               |    8             8
      solve_affine_greaterthan            |    8             8
      solve_affine_equalto                |    8             8
      solve_affine_interval               |    8             8
      solve_duplicate_terms_scalar_affine |    8             8
      solve_duplicate_terms_vector_affine |    8             8
      solve_qcp_edge_cases                |   14            14
      solve_qp_edge_cases                 |   21            21
      solve_duplicate_terms_obj           |   10            10
      solve_affine_deletion_edge_cases    |   24            24
      solve_integer_edge_cases            |   20            20
      solve_objbound_edge_cases           |   24            24
      raw_status_string                   |    6             6
      solve_time                          |    6             6
      solve_zero_one_with_bounds          |    9             9
      solve_unbounded_model               |    1             1
    modifications                         |  137           137
  Continuous Linear                       |  737           737
  Continuous Conic                        |  965           965
  Continuous Quadratic                    |  227           227
  Integer Linear                          |  180           180
  Integer Conic                           |   20            20
ERROR: LoadError: Some tests did not pass: 3197 passed, 0 failed, 2 errored, 0 broken.
in expression starting at /.../MathOptInterface.jl/test/runtests.jl:31
ERROR: Package MathOptInterface errored during testing
```

Reply:
Same error occurs if I remove `MOI.set(mock, MOI.Silent(), true)`

Reply:
@blegat, any ideas?

Reply:
It's just because the mock optimizer doesn't implement the default values, so they need to be set here. I think it's okay to set to `nothing` here for now.

Reply:
Then why I don't add Silent and TimeLimitSec here:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/0923d1e3a5fc9fdabc9c322bee84f29abe3805c3/src/Utilities/mockoptimizer.jl#L89

That way I can set the default value to what they should be. Then we can add test checking for that values to be sure every solver implement them as default.

Reply:
Sure. That seems like the right thing to do.

Reply:
Yes, we could set it directly in the constructor for `Silent` and `TimeLimitSec`. You should also update `is_empty` so that it does not check `isempty(mock.optimizer_attribute)`

Reply:
I'm working on it, I changed `empty!()` too. Still `set` and `get` aren't working. They change `mock.inner_model` because `Silent` and `TimeLimitSec` aren't set by optimize:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/0923d1e3a5fc9fdabc9c322bee84f29abe3805c3/src/Utilities/mockoptimizer.jl#L197-L203

Reply:
In `function MockOptimizer()` should I leave `Dict{MOI.AbstractOptimizerAttribute, Any}(),` and modify `inner_model`?

Reply:
I did try to change `inner_model` when `MockOptimizer()` is called but since `Silent()` only works with `UniversalFallback` that fails.
I could modify the way `set` and `get` work so we can store `Silent()` and `TimeLimitSec()` in `mock` instead of `mock.inner_model`. I'll try that and if I can get anything working I'll push so you can review.

Reply:
I have a temporary workaround. I use a `try catch` in `set` and `get`.
The problem is that I didn't find a way to know if `inner_model` supports an `AbstractOptimizerAttribute`. If it does and I stores it in `optimizer_attributes` then `is_empty` fails:
```Julia
AssertionError: MOI.is_empty(optimizer)
  Stacktrace:
   [1] MathOptInterface.Utilities.CachingOptimizer(::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}, ::MathOptInterface.Utilities.MockOptimizer{MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}) at /.../MathOptInterface.jl/src/Utilities/cachingoptimizer.jl:61
...```
I tried to use `supports` but it's not defined since `Silent()` and `TimeLimitSec()` aren't copyable. Maybe I didn't understand what `supports` is used for.

Reply:
> I did try to change `inner_model` when `MockOptimizer()` is called but since `Silent()` only works with `UniversalFallback` that fails.

Why don't you use a `UniversalFallback` ?

Reply:
I think I have no control on this. If there is a test that don't use `UniversalFallback` `inner_model` can't accept `Silent()` and `TimeLimitSec()`. I think I need to `set` those attributes with default values without knowing which `ModelLike` is given to `MockOptimizer`.

Reply:
In `src/Test/unit.jl`, you cannot assume what the `ModelLike` is but in `test/Test/unit.Jl` you can choose to use a `UniversalFallback` like here:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/d3106f434293a2aae7c664a22a30a7bd4069111a/test/Test/unit.jl#L14-L15

Reply:
I will change that then but `MockOptimizer` will always be using `UniversalFallback`. It feels that it shouldn't be that way.

Reply:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/d3106f434293a2aae7c664a22a30a7bd4069111a/test/Utilities/copy.jl#L30-L39
That breaks if I try to change the `::ModelLike` for UniversalFallback.

Reply:
You only need to change it in `test/Test/unit.jl`, don't modify anything in `src`

Reply:
I still have to make changes in `src/Utilities/mockoptimizer.jl` to set default values. With that done if I don't change `set` and `get` this happen - `MOI.set(mock.inner_model, attr, value)` - on MockOptimizer call.
That's fine if the call is made with UniversalFallback so I'm replacing `MOIU.Model{Float64}()` with `MOIU.UniversalFallback(MOIU.Model{Float64}())` in `test/`.

When I did that in `test/Utilities/copy.jl` I got new errors:
```Julia
@testset "Allocate-Load" begin
    @test !MOIU.supports_allocate_load(DummyModel(), false)
    @test !MOIU.supports_allocate_load(DummyModel(), true)
    mock = MOIU.MockOptimizer(MOIU.UniversalFallback(MOIU.Model{Float64}()), needs_allocate_load=true)
    MOIT.failcopytestc(mock)
    MOIT.failcopytestia(mock)
    MOIT.failcopytestva(mock)
    MOIT.failcopytestca(mock)
    MOIT.copytest(mock, MOIU.UniversalFallback(MOIU.Model{Float64}()))
end
```
```Julia
Test Summary:        | Pass  Fail  Error  Total
MOI.Utilities        | 5182     8      1   5191
  Functions          |  388                 388
  Sets               |   63                  63
  Constraints        |    8                   8
  Model              |  948                 948
  Universal Fallback |  661                 661
  Parser             |   35                  35
  Mock Optimizer     |  203                 203
  Caching Optimizer  | 2756                2756
  Copy               |   58     8      1     67
    AUTOMATIC        |    2                   2
    Default          |   47                  47
    Allocate-Load    |    9     8      1     18
  CleverDicts        |   62                  62
```

Reply:
> I still have to make changes in `src/Utilities/mockoptimizer.jl` to set default values

Indeed, this requires the internal model to always support `TimeLimitSec`. You can either not set the default (like it is currently done for `Silent`) or only set it in case the internal model supports it (you can check that with `supports`).
I would rather do the first option (not setting the default) as the other behavior is somewhat weird, setting the optimizer attributes of ModelLike passed to MockOptimizer is a rather suprising behavior

Reply:
I think that would be strange to put `default value is` in docs but not testing for it because it's not implemented in MOI side.
I had issue with the use of `supports` since it returns an error instead of false for `Silent()` and `TimeLimitSec()`:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/ab2d0f05205a81c3848f9f1d870b75332b1bddd2/src/attributes.jl#L155-L169

To be honest I don't know how to solve that without some important changes as this bad `try catch` I did in last push.

Reply:
> it's not implemented in MOI side

`MockOptimizer` is just a utility for testing so I don't think it is an issul if it does not have the default values for `Silent` and `TimeLimitSec`.
If new optimizer attributes are defined outside of MOI, `MockOptimizer` won't be able to have the right default values. It's case is a bit specific since it supports **any** optimizer attributes (when used with UniversalFallback), even some that may not defined in MOI. But it actually completely ignores their value.
That's not the case for solvers that only support specific attributes, take their value into account and should have the correct default values.

Reply:
My understanding is: if I remove default from `MockOptimizer` then MOI can't have a test to check if Optimizers implement defaults correctly. If I didn't understand that well then I don't see why `MockOptimizer` should have this feature but if I'm right then I believe this feature is quite important.

I'll push without the default values.

Reply:
> if I remove default from `MockOptimizer` then MOI can't have a test to check if Optimizers implement defaults correctly.

You can set the attribute's value to its default before running the tests as it is done for `Silent` here:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/42cdc3454cdb88345da0adb9cce2a57497c2fad4/test/Test/unit.jl#L15-L16
Of course solvers shouldn't do that but for the `MockOptimizer` it's ok to mock the test